design/configuration-crd.md

```diff
@@ -0,0 +1,250 @@
+timeouts:
+  request-timeout: infinity
+  connection-idle-timeout: 60s
+  stream-idle-timeout: 5m
+  max-connection-duration: infinity
+  delayed-close-timeout: 1s
+  connection-shutdown-grace-period: 5s
+cluster:
+  dns-lookup-family: auto
+network:
+  num-trusted-hops: 0
+  admin-port: 9001
+rateLimitService:
+  extensionService: projectcontour/ratelimit
+  domain: contour
+  failOpen: false
+  enableXRateLimitHeaders: false
+policy:
+  request-headers:
+    set:
+  response-headers:
+    set:
+```
+
+The contents of this file has grown over time and some fields need to be better categorized. 
+The new ContourConfiguration CRD will move the logging options into a `logging` struct. 
+All other fields will stay the same where they currently exist in the configmap.
+
+### New fields:
+
+Some fields will be moved around and others added to complete the configuration:
+- logging: The logging configuration bits are now moved into a common struct
+- envoy: This manages the Envoy configuration of how it should be deployed and exposed from the cluster
+- ingressClassName: Adds the `ingress-class-name` flag to the configuration file
+
+## New CRD Spec
+
+```yaml
+apiVersion: projectcontour.io/v1alpha1
+kind: ContourConfiguration
+metadata:
+  name: contour
+spec:
+  server:
+    xds-server-type: contour
+  ingressClassName: contour
+  envoy:  
+    networkPublishing:
+    nodePlacement:
+      nodeSelector:
+      tolerations:
+  gateway:
+      controllerName: projectcontour.io/projectcontour/contour
+  incluster: true
+  kubeconfig: /path/to/.kube/config
+  disableAllowChunkedLength: false
+  disablePermitInsecure: false
+  tls:
+    minimum-protocol-version: "1.2"
+    cipher-suites:
+      - '[ECDHE-ECDSA-AES128-GCM-SHA256|ECDHE-ECDSA-CHACHA20-POLY1305]'
+      - '[ECDHE-RSA-AES128-GCM-SHA256|ECDHE-RSA-CHACHA20-POLY1305]'
+      - 'ECDHE-ECDSA-AES256-GCM-SHA384'
+      - 'ECDHE-RSA-AES256-GCM-SHA384'
+  fallback-certificate:
+    name: fallback-secret-name
+    namespace: projectcontour
+  envoy-client-certificate:
+    name: envoy-client-cert-secret-name
+    namespace: projectcontour
+  leaderelection:
+    configmap-name: leader-elect
+    configmap-namespace: projectcontour
+  enableExternalNameService: false
+  logging: 
+    accesslog-format: envoy
+    accesslog-format-string: "...\n"
+    json-fields:
+      - <Fields Omitted)
+  default-http-versions:
+    - "HTTP/2"
+    - "HTTP/1.1"
+  timeouts:
+    request-timeout: infinity
+    connection-idle-timeout: 60s
+    stream-idle-timeout: 5m
+    max-connection-duration: infinity
+    delayed-close-timeout: 1s
+    connection-shutdown-grace-period: 5s
+  cluster:
+    dns-lookup-family: auto
+  network:
+    num-trusted-hops: 0
+    admin-port: 9001
+  rateLimitService:
+    extensionService: projectcontour/ratelimit
+    domain: contour
+    failOpen: false
+    enableXRateLimitHeaders: false
+  policy:
+    request-headers:
+      set:
+    response-headers:
+      set:
+status:
+```
+
```

Comment:
Yeah, a rough outline of these cases would be good, as well as the behavior for when the CRD is changed while Contour is running; if we intend for this to be fully editable on-the-fly, we need to make sure that it only contains things that can be edited on the fly, otherwise we will end up effectively wrapping the current `main()` in a "reconcile ContourConfiguration CRD loop`, which I don't think is much improvement on the manual restart process.